src/backtracking/graph_coloring.rs

```diff
@@ -0,0 +1,351 @@
+    num_colors: usize,
+) -> Result<Option<Vec<Vec<usize>>>, GraphColoringError> {
+    GraphColoring::new(adjacency_matrix)?.find_solutions(num_colors)
+}
+
+/// A struct representing a graph coloring problem.
+struct GraphColoring {
+    // The adjacency matrix of the graph
+    adjacency_matrix: Vec<Vec<bool>>,
+    // The current colors assigned to each vertex
+    vertex_colors: Vec<usize>,
+    // Vector of all valid color assignments for the vertices found during the search
+    solutions: Vec<Vec<usize>>,
+}
+
+impl GraphColoring {
+    /// Creates a new GraphColoring instance.
+    ///
+    /// # Arguments
+    ///
+    /// * `adjacency_matrix` - A 2D vector representing the adjacency matrix of the graph.
+    ///
+    /// # Returns
+    ///
+    /// * A new instance of GraphColoring or a `GraphColoringError` if the matrix is empty or non-square.
+    fn new(adjacency_matrix: Vec<Vec<bool>>) -> Result<Self, GraphColoringError> {
+        let num_vertices = adjacency_matrix.len();
+
+        // Check if the adjacency matrix is empty
+        if num_vertices == 0 {
+            return Err(GraphColoringError::EmptyAdjacencyMatrix);
+        }
+
+        // Check if the adjacency matrix is square
+        if adjacency_matrix.iter().any(|row| row.len() != num_vertices) {
+            return Err(GraphColoringError::ImproperAdjacencyMatrix);
+        }
+
+        Ok(GraphColoring {
+            adjacency_matrix,
+            vertex_colors: vec![0; num_vertices],
+            solutions: Vec::new(),
+        })
+    }
+
+    /// Returns the number of vertices in the graph.
+    fn num_vertices(&self) -> usize {
+        self.adjacency_matrix.len()
+    }
+
+    /// Checks if a given color can be assigned to a vertex without causing a conflict.
+    ///
+    /// # Arguments
+    ///
+    /// * `vertex` - The index of the vertex to be colored.
+    /// * `color` - The color to be assigned to the vertex.
+    ///
+    /// # Returns
+    ///
+    /// * `true` if the color can be assigned to the vertex, `false` otherwise.
+    fn is_color_valid(&self, vertex: usize, color: usize) -> bool {
+        for neighbor in 0..self.num_vertices() {
+            // Check outgoing edges from vertex and incoming edges to vertex
+            if (self.adjacency_matrix[vertex][neighbor] || self.adjacency_matrix[neighbor][vertex])
+                && self.vertex_colors[neighbor] == color
+            {
+                return false;
+            }
+        }
+        true
+    }
+
+    /// Recursively finds all valid colorings for the graph.
+    ///
+    /// # Arguments
+    ///
+    /// * `vertex` - The current vertex to be colored.
+    /// * `num_colors` - The number of colors available for coloring the graph.
+    fn find_colorings(&mut self, vertex: usize, num_colors: usize) {
+        if vertex == self.num_vertices() {
+            self.solutions.push(self.vertex_colors.clone());
+            return;
+        }
+
+        for color in 1..=num_colors {
```

Comment:
Is there a reason why you have chosen `1` to be the first color? It seams to be more natural (in languages like rust) to start with 0.

Reply:
There are no special reasons, I just use it on my own. After reviewing your suggestion, I see that it is more natural to start counting from 0, and using 0 as the first color can also make the implementation more consistent with typical array indexing.